support setting the clock manually via timeserver attestations

awwad opened the pull request:
See #173 for the description of the broader issue.

This PR supports the use of upTUF's manual time override (https://github.com/awwad/tuf/pull/22) to override system time on the Primary and Secondary client when time attestations are verified.

When a time attestation is verified, the clock that TUF and Uptane will use is set to that time, and will remain at that time until updated in the same way.

Tests already employ this functionality, and all added code is covered already, but separate unit (#176) and integration tests (#175) are desirable.  Those issues have been created to cover those needs.

awwad committed:
WIP: support setting the clock manually via timeserver attestations

This depends on the branch with the same name in awwad/tuf (Uptane's
TUF fork).  Note that when this is updated after that branch is
merged into develop on awwad/tuf, the dev-requirements.txt change
here should be removed.

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

awwad committed:
PR revision: correct time parsing in clients

Convert iso8601 time strings to unix times correctly,
in both primary.py and secondary.py.

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

awwad committed:
PR revision: move back to develop branch of uptuf

Now that the required PR is merged (uptuf awwad/tuf PR #22), we
can go back to using the develop branch in our dev-requirements.txt
file.

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

lukpueh commented on uptane/clients/primary.py:
```diff
@@ -1154,13 +1159,22 @@ def validate_time_attestation(self, timeserver_attestation):
+    # attestation and time.  Convert to a UNIX timestamp.
+    new_timeserver_time_unix = int(tuf.formats.datetime_to_unix_timestamp(
+        iso8601.parse_date(new_timeserver_time)))
+    tuf.formats.UNIX_TIMESTAMP_SCHEMA.check_match(new_timeserver_time_unix)
```
Stupid question but shouldn't `datetime_to_unix_timestamp` be guaranteed to return a unix timestamp (if it doesn't raise an exception)?

awwad replied:
It seems to always return a unix timestamp unless it raises an error, but it's still code sitting in a different project, so I felt like a defensive check might be wise.

lukpueh commented on uptane/clients/primary.py:
```diff
@@ -1154,13 +1159,22 @@ def validate_time_attestation(self, timeserver_attestation):
 
+    # Set the client's clock.  This will be used instead of system time by TUF.
+    tuf.conf.CLOCK_OVERRIDE = new_timeserver_time_unix
+
```
This seems like a serious side effect. Should it be mentioned in the function's docstring?

awwad replied:
I think that would be best, yes; thanks.  This is the current design in Uptane, though, a ratcheting "clock" based on verified timeserver attestations, so it should be unsurprising to users that we do this somewhere in the update cycle.  It's an important note to make in a function that claims to do validation, though.

awwad replied:
While it looks like the docstring already says "the Primary's time is updated and the attestation will be saved so that it can be provided to Secondaries.", I'm going to change the function name and write a better docstring.

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -440,9 +441,18 @@ def validate_time_attestation(self, timeserver_attestation):
 
+    # Set the client's clock.  This will be used instead of system time by TUF.
+    tuf.conf.CLOCK_OVERRIDE = new_timeserver_time_unix
+
```
Same comments as in `primary.py` apply here.

lukpueh reviewed (commented):
The diff *looks good to me*. But I fear I am not familiar enough to understand its full impact.

awwad committed:
Clarify language around verifying time attestations,

including:
- validate_time_attestation -> update_time
- validate -> verify (where it pertains to time attestations)
- note, in a few additional locations, that verified updated times
  will be used in place of the system time to check expiration on
  metadata.

The broader corrections of "validate" to "verify" (etc.) in the
codebase are left for another PR.

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

awwad committed:
Client tests for time override

plus:
- some replacements in test code of "validate" with "verify"
- some improved testing for time attestation verification and
  use

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

awwad commented:
I addressed @lukpueh's comments (refining docstrings -- and renaming a func for good measure).  Thanks, Lukas. 😄

I also added some testing and ran the demo again.  @lukpueh, if you have a second to look over those last two commits added after your review, that'd be appreciated, but I'm merging.

lukpueh commented on tests/test_primary.py:
```diff
@@ -578,30 +579,46 @@ def test_20_validate_time_attestation(self):
           private_key=TestPrimary.key_timeserver_pri, resign=True)
 
+    # Check expected base conditions before updating time:
+    # The only timeserver times registered should be two "now"s added during
```
Could it be that this is a copy-paste mistake (from `test_secondary.py`)? You say that it should have "two" registered times but are asserting for `len==1` below. Maybe I'm missing something.

awwad replied:
Fixed both now, thanks. :)

lukpueh reviewed (commented):
LGTM! `update_time` IMO is a better fit, and the new comments/docs also seem sensible. I did a quick grep for the old function name and I think you [missed one replace in a test comment](https://github.com/uptane/uptane/blob/dfd37b2c921a872b471e5d01b8b2421a4dd190ed/tests/test_primary.py#L909): `s/test_20_validate_time_attestation/test_20_update_time`
